src/gui/sdlmain.cpp

```diff
@@ -1978,107 +1976,178 @@ static bool detect_resizable_window()
+static SDL_Point remove_43_aspect(const SDL_Point &size)
 {
-#if C_OPENGL
-	if (sdl.desktop.want_type != SCREEN_OPENGL)
-		return false;
+	return {size.x, ceil_sdivide(size.y * 5, 6)};
+}
 
-	const std::string sname = get_glshader_value();
-	return (sname == "crt-fakelottes-flat" || sname == "crt-easymode-flat");
-#else
-	return false;
-#endif // C_OPENGL
+static SDL_Point refine_window_size_with_mode(const SDL_Point &size,
+                                              const SCALING_MODE scaling_mode,
+                                              const bool wants_43_aspect)
+{
+	switch (scaling_mode) {
+	case (SCALING_MODE::NONE): {
+		const auto game_ratios = wants_43_aspect ? SDL_Point{4, 3}
+		                                         : SDL_Point{8, 5};
+		const auto game_aspect = static_cast<double>(game_ratios.x) /
+		                         game_ratios.y;
+		const auto window_aspect = static_cast<double>(size.x) / size.y;
+		if (window_aspect > game_aspect) {
+			const int x = ceil_sdivide(size.y * game_ratios.x,
+			                           game_ratios.y);
+			return {x, size.y};
+		} else { // vice-versa
+			const int y = ceil_sdivide(size.x * game_ratios.y,
+			                           game_ratios.x);
+			return {size.x, y};
+		}
+	}
+	case (SCALING_MODE::NEAREST): {
+		constexpr SDL_Point resolutions[] = {
+		        {7680, 5760}, // 8K  at 4:3 aspect
+		        {7360, 5520}, //
+		        {7040, 5280}, //
+		        {6720, 5040}, //
+		        {6400, 4800}, // HUXGA
+		        {6080, 4560}, //
+		        {5760, 4320}, // 8K "Full Format" at 4:3 aspect
+		        {5440, 4080}, //
+		        {5120, 3840}, // HSXGA at 4:3 aspect
+		        {4800, 3600}, //
+		        {4480, 3360}, //
+		        {4160, 3120}, //
+		        {3840, 2880}, // 4K UHD at 4:3 aspect
+		        {3520, 2640}, //
+		        {3200, 2400}, // QUXGA
+		        {2880, 2160}, // 3K UHD at 4:3 aspect
+		        {2560, 1920}, // 4.92M3 (Max CRT, Viewsonic P225f)
+		        {2400, 1800}, //
+		        {1920, 1440}, // 1080p in 4:3
+		        {1600, 1200}, // UXGA
+		        {1280, 960},  // 720p in 4:3
+		        {1024, 768},  // XGA
+		        {800, 600},   // SVGA
+		};
+		// Pick the biggest window size that fits inside the bounds.
+		for (const auto &candidate : resolutions)
+			if (candidate.x <= size.x && candidate.y <= size.y)
+				return wants_43_aspect
+				               ? candidate
+				               : remove_43_aspect(candidate);
+		break;
+	}
+	case (SCALING_MODE::PERFECT): {
+		int x = 0;
+		int y = 0;
+		constexpr double aspect_weight = 1.14;
+		const double pixel_aspect_ratio = wants_43_aspect ? 1.2 : 1;
+		const int err = pp_getscale(320, 200, pixel_aspect_ratio, size.x,
+		                            size.y, aspect_weight, &x, &y);
+		if (err == 0)
+			return {320 * x, 200 * y};
+		break;
+	}
+	};
+	const SDL_Point fallback_size = {640, 480};
+	return wants_43_aspect ? fallback_size : remove_43_aspect(fallback_size);
```

Comment:
Suggestion - maybe the 640 x 480 fallback here and down on line 2077 should be a global `constexpr` variable? Also gets rid of magic numbers.